// @sudoku/stores/candidates.js
import { writable } from 'svelte/store';

function createCandidates() {
	const candidates = writable({});

	return {
		subscribe: candidates.subscribe,

		add(pos, candidate) {
			candidates.update($candidates => {
				if (!$candidates.hasOwnProperty(pos.x + ',' + pos.y)) {
					$candidates[pos.x + ',' + pos.y] = [candidate];
				} else if ($candidates[pos.x + ',' + pos.y].includes(candidate)) {
					delete $candidates[pos.x + ',' + pos.y][$candidates[pos.x + ',' + pos.y].indexOf(candidate)];
				} else {
					$candidates[pos.x + ',' + pos.y].push(candidate);
				}

				return $candidates;
			});
		},

		clear(pos) {
			candidates.update($candidates => {
				delete $candidates[pos.x + ',' + pos.y];
				return $candidates;
			});
		},

		set(pos, candidateList) {
			candidates.update($candidates => {
				const key = pos.x + ',' + pos.y;
				const normalized = Array.from(new Set((candidateList || []).filter(v => Number.isInteger(v) && v >= 1 && v <= 9))).sort((a, b) => a - b);

				if (normalized.length === 0) {
					delete $candidates[key];
				} else {
					$candidates[key] = normalized;
				}

				return $candidates;
			});
		},

		reset() {
			candidates.set({});
		}
	}
}

export const candidates = createCandidates();
